Asterix/optics/phase_amplitude_functions.py

```diff
@@ -3,52 +3,77 @@
 from Asterix.utils import crop_or_pad_image, rebin
 
 
-def roundpupil(dim_pp, prad, no_pixel=False, center_pos='b'):
-    """Create a circular pupil.
+def roundpupil(dim_pp, prad, grey_pup_bin_factor=1, center_pos='b'):
+    """
+    Create a circular pupil.
 
-    With no_pixel=True, this is a way to create a 10x oversampled pupil that is then rescaled to the requested size
-    using rebin().
+    With grey_pup_bin_factor >1, this create am oversized pupil that is then rescaled using rebin to dim_pp
 
     AUTHORS : Axel Pottier, Johan Mazoyer
     7/9/22 Modified by J Mazoyer to remove the pixel crenellation with rebin and add a better center option
+    2/10/22 Modified by J Mazoyer to enhance pixel grey_pup_bin_factor factor
 
     Parameters
     ----------
     dim_pp : int
-        Size of the image array (in pixels).
+        Size of the image (in pixels)
     prad : float
-        Pupil radius within the image array (in pixels).
-    no_pixel : boolean (default False).
-        If true, the pupil is first defined at a very large
-        scale (prad = 10*prad) and then rescaled to the given parameter 'prad'.
-        This limits the pixel crenellation in the pupil for small pupils.
+        Size of the pupil radius (in pixels)
     center_pos : string (optional, default 'b')
         Option for the center pixel.
         If 'p', center on the pixel dim_pp//2.
         If 'b', center in between pixels dim_pp//2 -1 and dim_pp//2, for 'dim_pp' odd or even.
+    grey_pup_bin_factor : int (default, 1)
+        If grey_pup_bin_factor > 1, the pupil is first defined at a very large scale
+        (prad = grey_pup_bin_factor*prad) and then rebined to the given parameter 'prad'.
+        This limits the pixel crenellation in the pupil for small pupils.
+        If this option is activated (grey_pup_bin_factor>1) the pupil has to be perfectly centered on
+        the array because binning and while keeping the centering is tricky:
+            -if center_pos is 'p', dimpp and grey_pup_bin_factor must both be odd
+            -if center_pos is 'b', dimpp and grey_pup_bin_factor must both be even
 
     Returns
     ------
     pupilnormal : 2D array
         Output circular pupil
     """
 
-    if no_pixel:
-        factor_bin = int(10)
-        pup_large = roundpupil(int(2 * prad) * factor_bin, factor_bin * prad, no_pixel=False)
-        return crop_or_pad_image(rebin(pup_large, factor=factor_bin, center_on_pixel=False), dim_pp)
+    if grey_pup_bin_factor > 1:
+        if grey_pup_bin_factor != int(grey_pup_bin_factor):
+            raise ValueError("grey_pup_bin_factor must be an integer")
```

Comment:
For clarity in the raised exception and its message, you could add: "[...], currently it is {grey_pup_bin_factor}" and make it an f-string.